### DISCUSSION TITLE
How get client ip

### DISCUSSION THREAD

**Discussion by ste22g** [CoreWCF/CoreWCF#1107] [Q&A] [answered]
Hi,
congratulations for the excellent work!

How can I get the ip of the service client?

With the 4.x .net version I used the following instructions:
```
 var clientIpAddress = System.Web.HttpContext.Current?.Request?.UserHostAddress;
 var hostIpPort = System.Web.HttpContext.Current?.Request?.Url?.Port;
```

Is there anyone who can give me information.
Thank you.

ste22g

**Comment by mconnew** [comment, marked as answer]
There's a couple of ways. In your service implementation, you can inject the HttpContext/HttpRequest/HttpResponse as a parameter. You don't need to modify your contract, just make your implementation partial (so CoreWCF code gen will backfill the now missing method), and add a new parameter to your implementation method attributed with `[CoreWCF.Injected]`.
E.g.
```c#
public class MyService : IMyService
{
    public void DoWork(string data)
    {
        // Your implementation    
    }
}
````
Becomes
```c#
public partial class MyService : IMyService
{
    public void DoWork(string data, [CoreWCF.Injected] HttpContext httpContext)
    {
        var clientIpAddress = httpContext.Connection.RemoteIpAddress;
        var clientPort = httpContext.Connection.RemotePort;
        // Your implementation    
    }
}
```

**Reply by ste22g** [reply]
Wow, fast and accurate answer.
Thank you.
ste22g

**Comment by igor-nesterov-deltatre** [comment]
What's the purpose of injecting it with partial class and  `[CoreWCF.Injected]` attribute, just to do not break the contract?

**Reply by mconnew** [reply]
Yes, that's exactly it. It enables you to say what stuff should be coming over the wire and what stuff is coming locally from DI.

**Comment by theburningfireeternal** [comment]
@mconnew how can i get the exact client ip, when the service in behind a load balancer ?

**Reply by mconnew** [reply]
Presuming HTTP, a lot of load balancers add the HTTP header `X-Forwarded-For` with the original client IP address. Mozilla has some useful docs here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For

Make sure you account for multiple hops. It can be a comma separated list and the original client IP should be the first item. You can get the headers easiest by making your class partial, then add a parameter to your implementation (leave the contract interface alone) that looks like `[CoreWCF.Injected]HttpRequest httpRequest` which will get you the incoming HttpRequest. You can then check the headers from there.

**Reply by g7ed6e** [reply]
@saurabh305 please refer to [ASP.NET core forward header middleware documentation](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-8.0) to configure how http headers are read and mapped to ASP.NET core object model when running behind a load balancer / reverse proxy.
